src/corebluetooth/peripheral.rs

```diff
@@ -76,17 +79,40 @@ impl Peripheral {
                     error!("Event receiver died, breaking out of corebluetooth device loop.");
                     break;
                 }
-                if let Some(CBPeripheralEvent::Notification(uuid, data)) = event {
-                    util::invoke_handlers(
-                        &nh_clone,
-                        &ValueNotification {
-                            uuid,
-                            handle: None,
-                            value: data,
-                        },
-                    );
-                } else {
-                    error!("Unhandled CBPeripheralEvent");
+                match event {
+                    Some(CBPeripheralEvent::Notification(uuid, data)) => {
+                        util::invoke_handlers(
+                            &nh_clone,
+                            &ValueNotification {
+                                uuid,
+                                handle: None,
+                                value: data,
+                            },
+                        );
+                    },
+                    Some(CBPeripheralEvent::ManufacturerData(manufacturer_id, data)) => {
+                        let mut properties = p_clone.lock().unwrap();
+                        properties.manufacturer_data.insert(manufacturer_id, data.clone());
+                        // TODO: emit event
+                        m_clone.emit(CentralEvent::ManufacturerAdvertisement{
+                            address: properties.address,
+                            manufacturer_id,
+                            data,
+                        });
+                    },
+                    Some(CBPeripheralEvent::ServiceData(service_data)) => {
+                        let mut properties = p_clone.lock().unwrap();
+                        properties.service_data.extend(service_data.clone());
```

Comment:
This could be bad as the service_data doesn't get cleared out, so over time, this may use up a bit of memory. I'm not sure if the service_data reports from Core Bluetooth are complete, or one off advertisements and whether the application wants to have access to older advertisement data

Reply:
The next question is, in regards to the API design, do we want to retain advertisement data, or emit them and forget?

Reply:
I think it's fine as the service UUIDs will be the same, or at least a small set, so they will be replaced in the HashMap won't they?

Reply:
According to [`impl<K, V, S> Extend<(K, V)> for HashMap<K, V, S>`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#impl-Extend%3C(K%2C%20V)%3E)
> Inserts all new key-values from the iterator and replaces values with existing keys with new values returned from the iterator.

Reply:
Right. So if the UUIDs are the same, which they generally will be, then the map won't get any bigger.